### PR TITLE
GSK-1943 - make jobs cancellable from the Hub

### PR DESCRIPTION
**Pull request by andreybavt**
main PR - https://github.com/Giskard-AI/giskard-hub/pull/289

### CODE REVIEW

File: giskard/commands/cli_worker.py [hunk 1]
```diff
@@ -160,7 +160,7 @@ def _start_command(is_server, url: AnyHttpUrl, api_key, is_daemon, hf_token=None
 
             run_daemon(is_server, url, api_key, hf_token)
         else:
-            if sys.platform == "win32":
+            if sys.platform == "win32" or sys.version_info < (3, 10):
```
Comment (andreybavt) [inline giskard/commands/cli_worker.py:163]: uvloop didn't work on my my with python 3.9, but it's technically not related to the feature, so can remove for now

File: giskard/commands/cli_worker.py [hunk 2]
```diff
@@ -160,9 +160,11 @@ def _start_command(is_server, url: AnyHttpUrl, api_key, is_daemon, hf_token=None
 
             run_daemon(is_server, url, api_key, hf_token)
         else:
-            if sys.platform == "win32":
+            if settings.force_asyncio_event_loop or sys.platform == "win32":
```
Comment (andreybavt) [inline giskard/commands/cli_worker.py:163]: I have issues when debugging worker from intellij when uvloop is used. I added the flag mostly for myself, but in case the same problems appear in prod there'll be a way to let people switch from one implementation to the other

File: giskard/ml_worker/websocket/listener.py [hunk 3]
```diff
@@ -221,10 +224,15 @@ async def wrapped_callback(req: ActionPayload, client_params: Dict, worker_info,
     return websocket_actor_callback
 
 
+@websocket_actor(MLWorkerAction.abort, execute_in_pool=False)
+def on_abort(ml_worker: MLWorkerInfo, params: AbortParams, *args, **kwargs):
+    POOL.cancel(params.job_id)
```
Comment (Inokinoki) [inline giskard/ml_worker/websocket/listener.py:229]: What if the job does not exist

**Review by Inokinoki — COMMENTED** [review]
LGTM

Will do some tests

File: giskard/commands/cli_worker.py (hunk 2 above)
Comment (kevinmessiaen) [inline giskard/commands/cli_worker.py:163]: Can be simplified in `elif`

File: giskard/ml_worker/websocket/listener.py [hunk 4]
```diff
@@ -134,6 +138,7 @@ def parse_and_execute(
     params,
     ml_worker: MLWorkerInfo,
     client_params: Optional[Dict[str, str]],
+    job_id: Optional[UUID],
```
Comment (kevinmessiaen) [inline giskard/ml_worker/websocket/listener.py:141]: Seems that we never use this param

File: giskard/ml_worker/websocket/listener.py (hunk 3 above)
Comment (andreybavt) [reply giskard/ml_worker/websocket/listener.py:229]: There's a check inside `WorkerPoolExecutor.kill_task`:
```
if future is not None and not future.cancel():
```
giskard/utils/worker_pool.py:272


so nothing will  be done in case the job doesn't exist, no error either

File: giskard/commands/cli_worker.py (hunk 2 above)
Comment (andreybavt) [reply giskard/commands/cli_worker.py:163]: hmm, not sure since there's a common `run(...)` on line 169 that should be executed in bot "else" subbranches.

File: giskard/ml_worker/websocket/listener.py [hunk 5]
```diff
@@ -100,6 +102,8 @@ async def handle_result(future: Union[Future, Callable[..., websocket.WorkerRepl
                 error_str=e.info.message, error_type=e.info.type, detail=e.info.stack_trace
             )
             logger.warning(e)
+        except TimeoutError as e:
```
Comment (Hartorn) [inline giskard/ml_worker/websocket/listener.py:105]: When is this happening vs the cancelled one ?

File: giskard/ml_worker/websocket/listener.py [hunk 6]
```diff
@@ -221,10 +223,15 @@ async def wrapped_callback(req: ActionPayload, client_params: Dict, worker_info,
     return websocket_actor_callback
 
 
+@websocket_actor(MLWorkerAction.abort, execute_in_pool=False)
+def on_abort(ml_worker: MLWorkerInfo, params: AbortParams, *args, **kwargs):
+    POOL.cancel(params.job_id)
```
Comment (Hartorn) [inline giskard/ml_worker/websocket/listener.py:228]: This should be wrapped, to hide the pool (same as for run_in_pool), for a matter of uniformity

Comment (Hartorn) [reply giskard/ml_worker/websocket/listener.py:228]: so you should just call a `cancel_in_pool(job_id)`

File: giskard/ml_worker/websocket/listener.py [hunk 7]
```diff
@@ -584,8 +591,12 @@ def run_test_suite(
 
 
 @websocket_actor(MLWorkerAction.echo, execute_in_pool=False)
-def echo(params: websocket.EchoMsg, *args, **kwargs) -> websocket.EchoMsg:
-    return params
+def echo(params: websocket.EchoMsg, *args, **kwargs) -> websocket.EchoResponse:
+    if POOL.pool:
```
Comment (Hartorn) [inline giskard/ml_worker/websocket/listener.py:595]: This should be handled in utils file also, you should just call `list_in_pool` or something

File: giskard/ml_worker/websocket/utils.py [hunk 8]
```diff
@@ -34,12 +34,15 @@
...
 from giskard.path_utils import projects_dir
+from giskard.ml_worker.websocket import AbortParams
 
 logger = logging.getLogger(__name__)
 
 
 def parse_action_param(action: MLWorkerAction, params):
     # TODO: Sort by usage frequency from future MixPanel metrics #NOSONAR
+    if action == MLWorkerAction.abort:
+        return AbortParams.model_validate(params)
```
Comment (Hartorn) [inline giskard/ml_worker/websocket/utils.py:45]: Is this compatible with pyantic 1 ? I think it would be better to stay with parse_obj legacy method for now

File: giskard/utils/__init__.py [hunk 9]
```diff
@@ -93,7 +99,7 @@ def pooled(fn):
 
     @wraps(fn)
     def wrapper(*args, **kwargs):
```
Comment (Hartorn) [inline giskard/utils/__init__.py:101]: I think we are not using this decorator, we may just remove it altogether.

File: giskard/utils/worker_pool.py [hunk 10]
```diff
@@ -74,16 +70,21 @@ def __init__(self) -> None:
...
-@dataclass(frozen=True)
 class GiskardTask:
-    callable: Callable
+    fn: Callable
     args: Any
     kwargs: Any
-    id: str = field(default_factory=_generate_task_id)
+    job_id: UUID
+
+    def __init__(self, job_id: UUID, fn: Callable, args: Any, kwargs: Any):
```
Comment (Hartorn) [inline giskard/utils/worker_pool.py:83]: This should be remove, since dataclass will define an implicit constructor.
I guess job_id should be moved as first attribute to respect order.

File: giskard/utils/worker_pool.py [hunk 11]
```diff
@@ -259,31 +266,32 @@ def spawn_worker(self):
...
+    def kill_task(self, job_id: UUID, reason: KillReason) -> Optional[BaseException]:
         # Task has timed out, we should kill it
         try:
-            future = self.futures_mapping.pop(task_id, None)
+            future = self.futures_mapping.pop(job_id, None)
             if future is not None and not future.cancel():
-                LOGGER.warning("Killing a timed out process")
-                future.set_exception(TimeoutError("Task took too long"))
-                pid = self.running_process.pop(task_id, None)
+                LOGGER.warning(f"Killing a job {job_id} with reason: {reason.name}")
```
Comment (Hartorn) [inline giskard/utils/worker_pool.py:274]: Side note, but it's better to not use f string for logging, but %s arguments, since computation would be skipped depending on level

File: tests/communications/test_dto_serialization.py [hunk 12]
```diff
@@ -85,9 +84,13 @@
     "TestSuiteParam": ["projectKey"],
     "WeightsPerFeature": [],
     "WorkerReply": [],
+    "AbortParams": ["job_id"],
```
Comment (Hartorn) [inline tests/communications/test_dto_serialization.py:87]: Could you use the code at the bottom of this file to generate the changes ? 
Items are sorted by alphabetical order here I think

File: tests/communications/test_dto_serialization.py [hunk 13]
```diff
@@ -85,9 +84,13 @@
...
+    "EchoResponse": ["job_ids", "msg"],
 }
 
 OPTIONAL_FIELDS = {
+    "AbortParams": [],
```
Comment (Hartorn) [inline tests/communications/test_dto_serialization.py:92]: Same

File: tests/communications/test_websocket_actor.py [hunk 14]
```diff
@@ -30,8 +30,9 @@ def test_all_registered_websocket_actor():
 def test_websocket_actor_echo():
     msg = websocket.EchoMsg(msg="echo")
     reply = listener.echo(msg)
-    assert isinstance(reply, websocket.EchoMsg)
+    assert isinstance(reply, websocket.EchoResponse)
     assert reply.msg == msg.msg
+    assert reply.job_ids == []
```
Comment (Hartorn) [inline tests/communications/test_websocket_actor.py:35]: You need to add tests for cancel task also

File: tests/llm/test_llm_client.py [hunk 15]
```diff
@@ -26,6 +27,7 @@
             finish_reason="stop",
             index=0,
             message=ChatCompletionMessage(content="This is a test!", role="assistant", function_call=None),
+            logprobs=None,
```
Comment (Hartorn) [inline tests/llm/test_llm_client.py:30]: unrelated change ?

File: pyproject.toml [hunk 16]
```diff
@@ -21,12 +21,6 @@ giskard = [
 ]
 
 
-[[tool.pdm.source]]
```
Comment (Hartorn) [inline pyproject.toml:24]: Unrelated changes ? or is it because branch is not up to date ?

**Review by Hartorn — CHANGES_REQUESTED** [review]

File: giskard/commands/cli_worker.py (hunk 2 above)
Comment (kevinmessiaen) [reply giskard/commands/cli_worker.py:163]: My bad I misread the indentation

File: giskard/ml_worker/websocket/listener.py (hunk 5 above)
Comment (andreybavt) [reply giskard/ml_worker/websocket/listener.py:105]: indeed, I tested, the timeout exceptions result in `CancelledError` so I don't see the need for a separate TimeoutError handler

File: giskard/ml_worker/websocket/listener.py (hunk 6 above)
Comment (andreybavt) [reply giskard/ml_worker/websocket/listener.py:228]: not sure I get it. Does the `cancel_in_pool` exist already?

Comment (andreybavt) [reply giskard/ml_worker/websocket/listener.py:228]: Why do you want to hide the POOL object inside utils?

File: giskard/utils/worker_pool.py (hunk 11 above)
Comment (andreybavt) [reply giskard/utils/worker_pool.py:274]: surprisingly Sonar doesn't warn about it...

File: tests/communications/test_dto_serialization.py (hunk 12 above)
Comment (andreybavt) [reply tests/communications/test_dto_serialization.py:87]: done

File: tests/communications/test_dto_serialization.py (hunk 13 above)
Comment (andreybavt) [reply tests/communications/test_dto_serialization.py:92]: done

File: tests/llm/test_llm_client.py (hunk 15 above)
Comment (andreybavt) [reply tests/llm/test_llm_client.py:30]: IIRC pydantic was complaining while this parameter was missing...

File: pyproject.toml (hunk 16 above)
Comment (andreybavt) [reply pyproject.toml]: not related to this PR

File: giskard/ml_worker/websocket/listener.py (hunk 6 above)
Comment (andreybavt) [reply giskard/ml_worker/websocket/listener.py]: done

File: giskard/ml_worker/websocket/listener.py (hunk 7 above)
Comment (andreybavt) [reply giskard/ml_worker/websocket/listener.py]: done

File: tests/communications/test_websocket_actor.py (hunk 14 above)
Comment (andreybavt) [reply tests/communications/test_websocket_actor.py:37]: done

File: giskard/ml_worker/websocket/utils.py (hunk 8 above)
Comment (andreybavt) [reply giskard/ml_worker/websocket/utils.py]: done

File: giskard/utils/worker_pool.py (hunk 10 above)
Comment (andreybavt) [reply giskard/utils/worker_pool.py]: done

**Review by Hartorn — APPROVED** [review]

**Review by kevinmessiaen — APPROVED** [review]
LGTM
